sympy/polys/polyutils.py

```diff
@@ -421,6 +438,17 @@ def _dict_reorder(rep, gens, new_gens):
 
     return map(tuple, new_monoms), coeffs
 
+def is_complex(exprs):
+    """Checks whether the numerator,Denominator of an expression is Complex"""
+    complex_in_num_denom = [False, False]
+    if len(exprs)>0:
+        if 'I' in str(exprs[0]):
```

Comment:
This looks fragile...what if a variable has the name "I2"? Are you looking for something in a given form or something that has a non-zero imaginary component?

Reply:
Maybe you are just trying to target expressions `e` for which `pure_complex(e)` is not None:
```python
>>> from sympy.core.evalf import pure_complex
>>> pure_complex(3+I)
(3, 1)
>>> pure_complex(3)
```

Reply:
I am just trying to check if the expression has any ImaginaryUnit in it
```
>>> from sympy.core.evalf import pure_complex
>>> pure_complex((2 + I)*(3 - I))
```
I was not expecting None here.
Can you please suggest any alternate way

Reply:
All expressions that have `I` in them? Then `exprs[0].has(I)` will work. But not all expressions that have I in them actually have imaginary components, e.g. roots of some cubics are returned in terms of `I` but are real.

What is the problem with `pure_complex` -- if you get None it means the expression does not look like `a + I*b`.

Reply:
As discussed in #18205, There was inconsistency in the results because in the case of `cancel((2+I)*(2-I))` because PolificationFailed was raised https://github.com/sympy/sympy/blob/f1334bb2094a16bc5465dccbfac0acfbc7fb383f/sympy/polys/polytools.py#L4287 The cause of this was opt.gens being empty because in `def _parallel_dict_from_expr_no_gens(exprs, opt):` exprs that is passed is multiplied i.e 5 in case of (2 + I)*(2 - I)).
 So it is checking the actual expression and if it has `I` in it, `I` in added in gens https://github.com/sympy/sympy/blob/f1334bb2094a16bc5465dccbfac0acfbc7fb383f/sympy/polys/polyutils.py#L286 As you can see the coefficient of added term is 0. So in case of cubics roots won't get affected.

There's no problem with `pure_complex`, its just that I don't understand how it can be implemented in this case.

Reply:
ping @smichr